### PR TITLE
Add tests for search top anomaly results API

### PR DESCRIPTION
**Pull request by ohltyler**
### Description
Add unit and integ tests for search top anomaly results API. Meets the jacoco coverage bar for all classes added as part of the API, removed from the exclusion list in `build.gradle`.
 

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by amitgalitz** [comment]
Are you planning on adding any integration tests with security

File: src/test/java/org/opensearch/ad/model/AnomalyResultBucketTests.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+package org.opensearch.ad.model;
+
+import java.io.IOException;
+
+import org.opensearch.ad.AbstractADTest;
+import org.opensearch.ad.TestHelpers;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.common.io.stream.StreamInput;
+
+public class AnomalyResultBucketTests extends AbstractADTest {
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/model/AnomalyResultBucketTests.java:15]: Could you add a test for the toXContent() method?

**Comment by ohltyler** [comment]
> Are you planning on adding any integration tests with security

Not as part of this PR. Seems there is a lot of missing API tests for security so seems it would be beneficial to do all of these together as a separate effort.

File: src/test/java/org/opensearch/ad/TestHelpers.java [hunk 2]
```diff
@@ -392,6 +387,61 @@ public static AnomalyDetector randomAnomalyDetectorUsingCategoryFields(String de
...
+            null,
+            randomInt(),
+            Instant.now(),
+            categoryFields,
+            randomUser(),
+            null
+        );
+    }
+
+    public static AnomalyDetector randomAnomalyDetectorUsingCategoryFields(
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/TestHelpers.java:417]: nit: multiple `randomAnomalyDetectorUsingCategoryFields` is a little confusing. Maybe can add w/ result index to this one or with custom indices to the one just above since that differentiates the methods. Also nit picky but these could be refactored into only calling the full `new AnomalyDetector(...)` once by gradually passing more variables like we do in `randomAnomalyDetector` if we choose to keep same name.

File: src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java [hunk 3]
```diff
@@ -1504,4 +1508,393 @@ public void testValidateAnomalyDetectorWithWrongCategoryField() throws Exception
...
+            true,
+            client()
+        );
+        Exception noCategoryFieldsException = expectThrows(
+            IOException.class,
+            () -> {
+                searchTopAnomalyResults(
+                    detectorWithNoCategoryFields.getDetectorId(),
+                    false,
+                    "{\"start_time_ms\":1, \"end_time_ms\":2}",
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1648]: nit: maybe we should pass category field here just to make it clear the issue is that the detector itself has no category field and not that search request is missing it.

File: src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java [hunk 4]
```diff
@@ -1504,4 +1508,393 @@ public void testValidateAnomalyDetectorWithWrongCategoryField() throws Exception
         );
 
     }
+
+    public void testSearchTopAnomalyResultsWithInvalidInputs() throws IOException {
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1512]: This is a lot of edge cases! Impressive!

File: src/test/java/org/opensearch/ad/TestHelpers.java [hunk 5]
```diff
@@ -1001,6 +1084,44 @@ public static void createIndex(RestClient client, String indexName, HttpEntity d
...
+        for (Map.Entry<String, String> entry : categoryFieldsAndTypes.entrySet()) {
+            indexMappings.append("\"" + entry.getKey() + "\":{\"type\":\"" + entry.getValue() + "\"},");
+        }
+        indexMappings.append("\"timestamp\":{\"type\":\"date\"}");
+        indexMappings.append("}}");
+        createEmptyIndex(client, indexName);
+        createIndexMapping(client, indexName, TestHelpers.toHttpEntity(indexMappings.toString()));
+    }
+
+    public static void createAnomalyResultIndex(RestClient client) throws IOException {
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/TestHelpers.java:1100]: nit: name change to `createEmptyAnomalyResultIndex` or `createAnomalyResultIndexWithNoData`

File: src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultActionTests.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            false,
+            1,
+            Arrays.asList(categoryFields.get(0)),
+            "invalid-order",
+            startTime,
+            endTime
+        );
+        expectThrows(
+            IllegalArgumentException.class,
+            () -> client().execute(SearchTopAnomalyResultAction.INSTANCE, searchRequest).actionGet(10_000)
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultActionTests.java:102]: smart to put timeout

File: src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultActionTests.java [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+import org.opensearch.ad.HistoricalAnalysisIntegTestCase;
+import org.opensearch.ad.TestHelpers;
+import org.opensearch.ad.model.AnomalyDetector;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.test.OpenSearchIntegTestCase;
+
+import com.google.common.collect.ImmutableList;
+
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.TEST)
+public class SearchTopAnomalyResultActionTests extends HistoricalAnalysisIntegTestCase {
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultActionTests.java:26]: Add a happy case also maybe?

File: src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java (hunk 3 above)
Comment (ohltyler) [reply src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1648]: The category field doesn't have to be passed in the request - the API will default to using all category fields found for the detector. I didn't want to pass a dummy category field value in the request since it would guarantee failure due to invalid category field, rather than missing category field from the detector itself. I tried to clarify the test case on line 1630.

File: src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultResponseTests.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        );
+
+        BytesStreamOutput output = new BytesStreamOutput();
+        originalResponse.writeTo(output);
+        StreamInput input = output.bytes().streamInput();
+        SearchTopAnomalyResultResponse parsedResponse = new SearchTopAnomalyResultResponse(input);
+        assertEquals(originalResponse.getAnomalyResultBuckets(), parsedResponse.getAnomalyResultBuckets());
+    }
+
+    public void testEmptyResults() {
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultResponseTests.java:31]: nit: I assume this is fine as is since there would be an exception or compiler error if you couldn't initialize an empty response but maybe add an assertEquals with expected empty response. Same for `testPopulatedResults` below.

File: src/test/java/org/opensearch/ad/TestHelpers.java (hunk 2 above)
Comment (ohltyler) [reply src/test/java/org/opensearch/ad/TestHelpers.java:417]: Good catch, will change to cascading constructors.

File: src/test/java/org/opensearch/ad/TestHelpers.java (hunk 5 above)
Comment (ohltyler) [reply src/test/java/org/opensearch/ad/TestHelpers.java:1100]: Sure, changed to first suggestion

File: src/test/java/org/opensearch/ad/model/AnomalyResultBucketTests.java (hunk 1 above)
Comment (ohltyler) [reply src/test/java/org/opensearch/ad/model/AnomalyResultBucketTests.java:15]: Sure, will add.

File: src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultActionTests.java (hunk 7 above)
Comment (ohltyler) [reply src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultActionTests.java:26]: Can't add happy cases due to issues with the `lang-painless` module not being installed on test clusters brought up with `OpenSearchIntegTest`s, so I cover those at the rest-level. I actually had a comment about that in there originally, I will add it back to prevent confusion.

Comment (ohltyler) [reply src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultActionTests.java:26]: See https://github.com/elastic/elasticsearch/issues/41109

File: src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultResponseTests.java (hunk 8 above)
Comment (ohltyler) [reply src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultResponseTests.java:31]: Prefer to not add an assertEquals here since it's a bit redundant. The response constructor that takes in a list (used here and in `testPopulatedResults`) simply assigns it to the single `anomalyResultBuckets` field, so not much to check.

File: src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java (hunk 3 above)
Comment (amitgalitz) [reply src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1648]: This makes sense to me! Thank you!

File: src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultActionTests.java (hunk 7 above)
Comment (amitgalitz) [reply src/test/java/org/opensearch/ad/transport/SearchTopAnomalyResultActionTests.java:26]: Okay I see, thanks for linking the issue. Comment could be nice to add as I just instinictaly thought of needing a happy case as I forgot about you mentioning this before.

File: src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java [hunk 9]
```diff
@@ -1504,4 +1508,393 @@ public void testValidateAnomalyDetectorWithWrongCategoryField() throws Exception
...
+                ),
+            true,
+            client()
+        );
+
+        // Clear any existing result index, create an empty one
+        if (indexExists(CommonName.ANOMALY_RESULT_INDEX_ALIAS)) {
+            deleteIndex(CommonName.ANOMALY_RESULT_INDEX_ALIAS);
+        }
+        TestHelpers.createEmptyAnomalyResultIndex(client());
```
Comment (ylwu-amzn) [inline src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1726]: Seems this is not the normal process from user's perspective. We don't allow user to create default AD result index directly. How about change to run historical analysis to create AD result index, then search top anomaly result?

Comment (ohltyler) [reply src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1726]: The purpose of this test was about having the index exist, but not return any results, so I just create the empty index so the API can query it, but still return empty results. Because this API doesn't interact with anything other than reading the results index, I still think this is ok. Same strategy is used in the below tests, as well as the existing search results tests.

I still prefer this than to run a historical analysis, so as to keep the test more lightweight, faster, and not increase the chances of it becoming flaky. Let me know if you still have concerns.

Comment (ylwu-amzn) [reply src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1726]: Got it. Sounds good. 
Another way to generate an empty default result index is run historical analysis on a short period (so it can run quickly), then delete all results with delete result REST API. But like you said, these steps will make the test heavier and more flaky likely.

**Review by amitgalitz — APPROVED** [review]

File: src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java [hunk 10]
```diff
@@ -1504,4 +1508,393 @@ public void testValidateAnomalyDetectorWithWrongCategoryField() throws Exception
         );
 
     }
+
+    public void testSearchTopAnomalyResultsWithInvalidInputs() throws IOException {
+        String indexName = randomAlphaOfLength(10).toLowerCase();
```
Comment (ylwu-amzn) [inline src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1513]: Index name must not contain some special chars like "#", "|", and must not start with "-", "+" or "_".
Not sure if the `randomAlphaOfLength` will always return valid index name. How about we use some concrete index name to avoid flaky test?

Comment (ylwu-amzn) [reply src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1513]: And same concern for detector name as Amit added regex checking detector name. @amitgalitz can you help check if we have such potential flaky part?

Comment (ohltyler) [reply src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1513]: Good catch. But, followed the stack of calls, and seems `randomAlphaOfLength` will only return `a-zA-Z`, since it's coming from this generator:
```
public class AsciiLettersGenerator extends CodepointSetGenerator {
  private final static char [] CHARS = 
      ("abcdefghijklmnopqrstuvwxyz" + 
       "ABCDEFGHIJKLMNOPQRSTUVWXYZ").toCharArray();

  public AsciiLettersGenerator() {
    super(CHARS);
  }
}
```

Comment (amitgalitz) [reply src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1513]: +1 on what Tyler mentioned, doesn't create anything flaky.

Comment (ylwu-amzn) [reply src/test/java/org/opensearch/ad/rest/AnomalyDetectorRestApiIT.java:1513]: Cool, thanks for confirming that

**Review by ylwu-amzn — APPROVED** [review]
